src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -309,25 +302,52 @@ private static string GetProjectOutputPath(string projectDirectory, string msbui
 
         internal static DependencyGraphSpec CreateMinimalDependencyGraphSpec(string projectPath, string outputPath)
         {
-            var packageSpec = new PackageSpec();
-            packageSpec.FilePath = projectPath;
-            packageSpec.RestoreMetadata = new ProjectRestoreMetadata();
-            packageSpec.RestoreMetadata.ProjectUniqueName = projectPath;
-            packageSpec.RestoreMetadata.ProjectStyle = ProjectStyle.PackageReference;
-            packageSpec.RestoreMetadata.ProjectPath = projectPath;
-            packageSpec.RestoreMetadata.OutputPath = outputPath;
-            packageSpec.RestoreMetadata.CacheFilePath = Path.Combine(outputPath, "project.nuget.cache");
+            var packageSpec = new PackageSpec
+            {
+                FilePath = projectPath,
+                RestoreMetadata = new ProjectRestoreMetadata()
+                {
+                    ProjectUniqueName = projectPath,
+                    ProjectStyle = ProjectStyle.PackageReference,
+                    ProjectPath = projectPath,
+                    OutputPath = outputPath,
+                    CacheFilePath = NoOpRestoreUtilities.GetProjectCacheFilePath(outputPath),
+                }
+            };
 
             var dgSpec = new DependencyGraphSpec();
             dgSpec.AddProject(packageSpec);
 
             return dgSpec;
         }
 
-        public Task RegisterRestoreInfoSourceAsync(IVsProjectRestoreInfoSource restoreInfoSource, CancellationToken cancellationToken)
+        public async Task RegisterRestoreInfoSourceAsync(IVsProjectRestoreInfoSource restoreInfoSource, CancellationToken token)
         {
-            return Task.CompletedTask;
-            // TODO: https://github.com/NuGet/Home/issues/10678
+            if (restoreInfoSource == null)
+            {
+                throw new ArgumentNullException(nameof(restoreInfoSource));
+            }
+
+            if (string.IsNullOrEmpty(restoreInfoSource.Name))
+            {
+                throw new ArgumentNullException(Resources.Argument_Cannot_Be_Null_Or_Empty, $"{nameof(restoreInfoSource)}.{nameof(restoreInfoSource.Name)}");
+            }
+
+            string projectUniqueName = restoreInfoSource.Name;
+            ProjectNames projectNames = await GetProjectNamesAsync(projectUniqueName, token);
+
+            _projectSystemCache.AddProjectRestoreInfoSource(projectNames, restoreInfoSource);
+        }
+
+        private async Task<ProjectNames> GetProjectNamesAsync(string projectUniqueName, CancellationToken token)
+        {
+            if (!_projectSystemCache.TryGetProjectNames(projectUniqueName, out ProjectNames projectNames))
+            {
+                IVsSolution2 vsSolution2 = await _vsSolution2.GetValueAsync(token);
+                projectNames = await ProjectNames.FromIVsSolution2(projectUniqueName, vsSolution2, token);
```

Comment:
logically, this requires UI thread, and might delay a project to be registered, so other projects might start a restore in-between, but maybe it can be rare in the real world to make it worth a fix.

Reply:
That scenario would normally happen during solution load correct? 
During solution load, NuGet checks that all projects have nominated anyways, so I think the likelihood of hitting this should be even lower.

Does that sound right?

Reply:
maybe, only when DTBB cache is ready for some projects might make it interesting. It is related to how HasPendingNomination/WhenNominated are called, but I cannot find how they are being used here. is this a complete PR, or I missed something?

Reply:
also, when net core calls register might make difference too.  currently it is unbounded (projects are doing its own, and no time limit when it is called, so if we run into race condition, it is more likely we need fix net core side first, instead of here.

Reply:
Yeah, we're not enforcing registration yet. 

Yes the PR is complete: https://github.com/NuGet/NuGet.Client/pull/4157/files#diff-10a8c14093035a8de36e9cec39e373608c246d4860742d1246290a11c926ed8a...`SolutionRestoreWorker` contains the change. GH might've collapsed it automatically in an effort to save vertical space.